### PR TITLE
web: Add a loading screen

### PR DESCRIPTION
**Pull request by Herschel**
Add a quick loading screen while the Ruffle WebAssembly loads.

This is a rough first pass loosely based on the custom loading bar on NG by @PsychoGoldfishNG. This really needs someone that is good with CSS to make it more functional and pretty -- I'm not that person! But I at least wanted to get this started.

TODO:
 * Make the loader fluid for very small containers, shrinking/hiding the Ruffle logo as the container shrinks.
 * Add a way to disable or slightly customize the preloader.
 * Get rid of the blue background. This works good for games, but it's not good for content like Flash web sites, which really want to display the native background color of the embed (#1752, #2135). Using a transparent BG is a bit ugly, though. Maybe we can use a smaller box instead of taking over the entire container.
 * Currently the progress bar only monitors the Wasm downloading, and not the SWF itself -- the progress bar will show as "full" while the SWF starts up. It's at least better than a white screen. We'd have to move the SWF fetch from Rust land to JavaScript to monitor its progress, but probably not worth it since we want to stream the SWF eventually.
 * Even when we have SWF streaming, we'll want to keep the preloader for the wasm download, as it's quite hefty and will only get bigger with time.

### CODE REVIEW

File: web/packages/core/src/load-ruffle.ts [hunk 1]
```diff
@@ -55,9 +61,40 @@ async function fetchRuffle(config: Config): Promise<typeof Ruffle> {
...
+        const bytesTotal = parseInt(contentLength, 10);
+        response = new Response(
+            new ReadableStream({
+                async start(controller) {
+                    const reader = wasmResponse.body?.getReader();
+                    if (!reader) {
+                        throw "Response had no body";
+                    }
+                    progressCallback?.(bytesLoaded, bytesTotal);
+                    for (;;) {
```
Comment (relrelb) [inline web/packages/core/src/load-ruffle.ts:81]: Looks like [`for await..of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of )can be used here.

File: web/packages/core/src/load-ruffle.ts [hunk 2]
```diff
@@ -55,9 +61,40 @@ async function fetchRuffle(config: Config): Promise<typeof Ruffle> {
     const { default: init, Ruffle } = await (extensionsSupported
         ? import("../pkg/ruffle_web-wasm_extensions")
         : import("../pkg/ruffle_web"));
+    let response;
+    const wasmUrl = extensionsSupported
+        ? new URL("../pkg/ruffle_web_bg.wasm", import.meta.url)
```
Comment (relrelb) [inline web/packages/core/src/load-ruffle.ts:66]: The URLs should be swapped.

**Comment by danielhjacobs** [comment]
> the progress bar will show as "full" while the SWF starts up

Does it? In my limited testing it seems to not show at all once the WASM has downloaded. A simple way to make it show (without consider for it staying at 100%) would be to move `this.isPreloaderVisible = false;` from the ensureFreshInstance function to the setMetadata function.

**Comment by danielhjacobs** [comment]
Tested with https://henrystickman.net/content/upload/games/files/fleeing-the-complex.swf

File: web/packages/core/src/shadow-template.ts [hunk 3]
```diff
@@ -227,3 +227,37 @@ ruffleShadowTemplate.innerHTML = `
...
+
+export const preloaderTemplate = document.createElement("template");
+preloaderTemplate.innerHTML = `
+    <div id="preloader">
+        <style>
+            #preloader {
+                display: flex;
+                flex-direction: column;
+                background: var(--ruffle-blue);
+                width: 100%;
```
Comment (danielhjacobs) [inline web/packages/core/src/shadow-template.ts:239]: width: 60%;

File: web/packages/core/src/shadow-template.ts [hunk 4]
```diff
@@ -227,3 +227,37 @@ ruffleShadowTemplate.innerHTML = `
...
+                height: 100%;
```
Comment (danielhjacobs) [inline web/packages/core/src/shadow-template.ts:240]: height: auto;

File: web/packages/core/src/shadow-template.ts [hunk 5]
```diff
@@ -227,3 +227,37 @@ ruffleShadowTemplate.innerHTML = `
...
+                padding: 10px auto;
+                align-items: center;
+                justify-content: center;
+            }
```
Comment (danielhjacobs) [inline web/packages/core/src/shadow-template.ts:244]: aspect-ratio: 2;
                position: absolute;
                top: 50%;
                left: 50%;
                transform: translateX(-50%) translateY(-50%);
    }

File: web/packages/core/src/shadow-template.ts [hunk 6]
```diff
@@ -227,3 +227,37 @@ ruffleShadowTemplate.innerHTML = `
...
 `;
+
+export const preloaderTemplate = document.createElement("template");
+preloaderTemplate.innerHTML = `
+    <div id="preloader">
+        <style>
+            #preloader {
+                display: flex;
+                flex-direction: column;
+                background: var(--ruffle-blue);
```
Comment (danielhjacobs) [inline web/packages/core/src/shadow-template.ts:238]: background: var(--preloader-background, var(--ruffle-blue));

That way, people who want to use a different background color can add this to the main site's CSS:

```
:root {
    --preloader-background: none; // or different color
}
```

File: web/packages/core/src/shadow-template.ts [hunk 7]
```diff
@@ -227,3 +227,37 @@ ruffleShadowTemplate.innerHTML = `
...
+                padding: 10px auto;
+                align-items: center;
+                justify-content: center;
+            }
+
+            .loadbar {
+                width: 100%;
+                max-width: 316px;
+                height: 10px;
+                margin-top: -15px;
```
Comment (danielhjacobs) [inline web/packages/core/src/shadow-template.ts:250]: margin-top: -3%;

File: web/packages/core/src/shadow-template.ts [hunk 8]
```diff
@@ -227,3 +227,37 @@ ruffleShadowTemplate.innerHTML = `
...
+                background: #253559;
+            }
+
+            .loadbarInner {
+                width: 0px;
+                height: 10px;
+                background: var(--ruffle-orange);
+            }
+        </style>
```
Comment (danielhjacobs) [inline web/packages/core/src/shadow-template.ts:259]: .logo {
            	display: var(--logo-display, block);
            }
        </style>
That way, people who don't want to use the Ruffle logo can add this to the main site's CSS:

```
:root {
    --logo-display: none;
}
```

File: web/packages/core/src/shadow-template.ts [hunk 9]
```diff
@@ -227,3 +227,37 @@ ruffleShadowTemplate.innerHTML = `
...
....4-.7.85-.25.45-.45.9-.15.45-.3.95l-5.85 41.6h16.25l5-35.5 1.5-11.25 4.2 6.3q.7 1.55 3.25 1.8l.05-.1q.25-.4.35-.85l.3-1.05 1.8-14.05v-.05l5.35-37.45h-16.25l-6.15 44.3 5.85 1.15h-9.45q-.5.05-1 .15-.5.15-1 .35-.5.2-.95.45m5.4-38.9q.15-1.7-.4-3.15-.5-1.1-1.35-2.1-2.25-2.25-5.5-2.25h-27.75q-2.3 0-4.45 1.35-.65.35-1.3.9-1.3.95-2.05 2.1-.45.6-.7 1.2-.4.9-.5 1.95l-4.15 29.95h-.05l-.7 5.2h-.05l-.2 1.35h.05l-.05.3 5.85 1.15h-9.45q-2.1.05-3.95 1.6-1.9 1.55-2.25 3.55l-.5 3.5h-.05l-1.2 8.75v.1l-4.1 29.25h16.25l5-35.5 1.5-11.25q.3 3.25 2.6 5.6.4.4.9.8 4.75 3.9 16 4.85l1.75-11.25h-12.2l.4-2.95h-.05l.7-5.05h-.05q.15-.9.3-1.9.1-.75.25-1.6.15-1.25.35-2.65v-.05q.95-6.7 2.35-16.5h11.25l-3.1 20.8h16.5l4.1-28.05M345 66.35h-.05l1.15-8.2q.5-3-1.75-5.25-1.25-1.25-3-1.75-1-.5-2.25-.5h-27.95q-.65 0-1.3.1-2.5.35-4.7 2.15-2.75 2.25-3.25 5.25l-1.95 14.7v.05l-.05.3 5.85 1.15h-9.45q-1.9.05-3.6 1.35-.2.1-.35.25-1.9 1.55-2.25 3.55l-4.85 34.1q-.25 3 1.75 5.25 1.25 1.4 3 1.95 1.05.3 2.25.3H320q3.25 0 6-2.25 2.75-2 3.25-5l2.75-18.5h-16.5l-1.75 11H302.5l2.1-14.75h.05l.85-6 1.5-11.2q.35 3.8 3.5 6.35 4.6 3.8 15.4 4.8.3 0 .6.05h15.75L345 66.35m-16.4-.95-1.25 8.95h-11.3l.4-2.95h-.05l.7-5.05h-.1l.15-.95h11.45Z"/></g></svg>
```
Comment (danielhjacobs) [inline web/packages/core/src/shadow-template.ts:260]: `<svg ...  "xMidYMid" viewBox="0 0 380 150" style="max-width: 380px; max-height: 150px; resize: horizontal; width: 100%; height: 100%;"><g><path fill="#966214" ... ><path fill="var(--ruffle-orange)" ... </g></svg>`

**Review by danielhjacobs — CHANGES_REQUESTED** [review]

File: web/packages/core/src/ruffle-player.ts [hunk 10]
```diff
@@ -455,6 +457,7 @@ export class RufflePlayer extends HTMLElement {
             this.container.style.visibility = "";
         }
 
+        this.isPreloaderVisible = false;
```
Comment (danielhjacobs) [inline web/packages/core/src/ruffle-player.ts:460]: Move this here: https://github.com/ruffle-rs/ruffle/blob/4f05d1852dd6c8e654aead0dcff7fefbc07264c9/web/packages/core/src/ruffle-player.ts#L1407

**Comment by Herschel** [comment]
Superseded by #7292.

**Pull request closed without merging**